### PR TITLE
keyhive sofle rgb: fix configuration.

### PR DESCRIPTION
**Pull request by winder**
## Description

Adds full support for the Keyhive Sofle RGB keyboard. The previous version was an incomplete mixture of changes from [SolarTempest](https://github.com/solartempest/qmk_firmware/) and [keyhive/qmk_firmware](https://github.com/keyhive/qmk_firmware). These changes fix via support by including the correct layout file, and extend it with support for rotary encoder mapping.

The readme files have been updated to reflect actual functionality.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by winder** [comment]
Updated to merge with the breaking changes!

**Comment by winder** [comment]
@SolarTempest would you mind reviewing this code? For the most part, it's a subset of the code in your repo.

File: keyboards/sofle/keyhive/config.h [hunk 1]
```diff
@@ -25,22 +25,22 @@
...
 #define MATRIX_ROWS  10
-#define MATRIX_COLS  6
+#define MATRIX_COLS  7
 
 // wiring of each half
 #define MATRIX_ROW_PINS \
     { C6, D7, E6, B4, B5 }
 #define MATRIX_COL_PINS \
-    { B6, B2, B3, B1, F7, F6 }
+    { B6, B2, B3, B1, F7, F6, C7 } //C7 is unused on the MCU. It is needed for the encoder key matrix in via.
```
Comment (drashna) [inline keyboards/sofle/keyhive/config.h:39]:
```suggestion
    { B6, B2, B3, B1, F7, F6, NO_PIN } //C7 is unused on the MCU. It is needed for the encoder key matrix in via.
```

File: keyboards/sofle/keyhive/config.h [hunk 2]
```diff
@@ -25,22 +25,22 @@
...
 #define MATRIX_ROW_PINS_RIGHT \
     { C6, D7, E6, B4, B5 }
 #define MATRIX_COL_PINS_RIGHT \
-    { F6, F7, B1, B3, B2, B6 }
+    { F6, F7, B1, B3, B2, B6, C7 } //C7 is unused on the MCU. It is needed for the encoder key matrix in via.
```
Comment (drashna) [inline keyboards/sofle/keyhive/config.h:43]:
```suggestion
    { F6, F7, B1, B3, B2, B6, NO_PIN } //C7 is unused on the MCU. It is needed for the encoder key matrix in via.
```

File: keyboards/sofle/keymaps/keyhive_via/config.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+/* The way how "handedness" is decided (which half is which),
+see https://docs.qmk.fm/#/feature_split_keyboard?id=setting-handedness
+for more options.
+*/
+
+#define RGB_DI_PIN D3
+#define RGBLED_NUM 74
+#define RGBLED_SPLIT { 37, 37 }
+#define RGBLIGHT_ANIMATIONS
```
Comment (drashna) [inline keyboards/sofle/keymaps/keyhive_via/config.h:28]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/sofle/keymaps/keyhive_via/via_layout.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (drashna) [inline keyboards/sofle/keymaps/keyhive_via/via_layout.json:1]: This is a VIA json, and belongs to the `the-via` repo, not here.  This needs to be removed.

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by winder** [comment]
@drashna thank you for the review! The problems you found should be addressed now.

**Comment by solartempest** [comment]
Took a look and just a couple comments:
1. Can you credit me with the readme keyboard photo?
2. Consider forcing NKRO if that is the intention: `#define FORCE_NKRO`.
3. I tried enabling RGB_MATRIX with WS2812 but it didn't work for me? Not sure if it was just what I was trying. RGB_LIGHT worked fine.

**Comment by winder** [comment]
Thanks @solartempest, let me know if the credit change is what you had in mind.

> 3. I tried enabling RGB_MATRIX with WS2812 but it didn't work for me? Not sure if it was just what I was trying. RGB_LIGHT worked fine.

Could you explain this more? I'm not sure what `RGB_MATRIX` / `RGB_LIGHT` are for. The LEDs seem to work fine

**Comment by solartempest** [comment]
I'm good with the credit, thanks!

Regarding RGB_MATRIX, I saw you changed some of the code here and had tried to test it but was unsuccessful: [https://github.com/qmk/qmk_firmware/pull/15339/files#diff-8a86f0d098cbbedffbc37a3dbf6550e60472ff8f0e68834b9a106219fb256f2dR95-R100](url)

In the QMK docs there are added details - RGB_MATRIX is used more for reactive lighting, while RGBLIGHT often used when that is not needed: https://docs.qmk.fm/#/feature_rgb_matrix . By default, the VIA keymap is using RGBLIGHT so those lines wouldn't actually need to be changed unless you wanted to (since you are not using that feature anyway).

Hope that helps clarify.

**Comment by winder** [comment]
@solartempest I think one of my rebases broke the link you shared. I wasn't trying to change anything with the rgb configuration, are you talking about LAYOUT_via? I added that so the other keymaps would work with sofle/keyhive

**Comment by solartempest** [comment]
> I think I see what you were talking about. This was part of the original code that I'm updating. I only noticed because some of the animations looked wrong when transitioning between halves. Your configuration looks much better so I ported that RGB config over along with everything else.

I see you have removed the RGB Matrix code. Looks much better and cleaned up! Only one typo for a missing 'k' in the readme.

File: keyboards/sofle/keyhive/readme.md [hunk 5]
```diff
@@ -1,66 +1,53 @@
...
+### Acknowledgements
 
-### Rotary Encoder and VIA Features:
--   Fully emappable left encoder controls volume up/down/mute. Right encoder PGUP/PGDOWN.
--   Allows for  live remapping of per-layer rotary encoder functions in VIA.
--   VIA support included by default.
--   This fixes the Keyhive left bottom row offset issue in VIA. You will need to import sofle.json in VIA.
--   Custom macro key in VIA for Super Alt Tab, which is fully compatible with rotary encoders.
--   Custom macro key in VIA for moving windows to other monitors in Windows, which is fully compatible with rotary encoders.
+* Solartempest - the image on this page and most of the code is either copied directly or inspired by their for. [Solartempest's fork.](https://github.com/solartempest/qmk_firmware/tree/master/keyboards/solartempest/sofle).
```
Comment (winder) [inline keyboards/sofle/keyhive/readme.md:17]:
```suggestion
* Solartempest - the image on this page and most of the code is either copied directly or inspired by their fork. [Solartempest's fork.](https://github.com/solartempest/qmk_firmware/tree/master/keyboards/solartempest/sofle).
```

File: keyboards/sofle/keymaps/via/keymap.c [hunk 6]
```diff
@@ -58,7 +58,7 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
  * `-----------------------------------------/       /     \      \-----------------------------------------'
  *            | LGUI | LAlt | LCTR |LOWER | /Enter  /       \Space \  |RAISE | RCTR | RAlt | RGUI |
  *            |      |      |      |      |/       /         \      \ |      |      |      |      |
- *            `----------------------------------'           '------''---------------------------'
+ *            `------------------------------_----'           '------''---------------------------'
```
Comment (fauxpark) [inline keyboards/sofle/keymaps/via/keymap.c:61]:
```suggestion
 *            `-----------------------------------'           '------''---------------------------'
```
?

File: keyboards/sofle/keyhive/rules.mk [hunk 7]
```diff
@@ -15,7 +15,7 @@ COMMAND_ENABLE = no         # Commands for debug and configuration
 # Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
 SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
 # if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
-NKRO_ENABLE = yes           # USB Nkey Rollover
+NKRO_ENABLE = no            # USB Nkey Rollover. 6KRO by default. NKRO works if enabled. (304)
```
Comment (fauxpark) [inline keyboards/sofle/keyhive/rules.mk:18]:
```suggestion
NKRO_ENABLE = no            # Enable N-Key Rollover
```

File: keyboards/sofle/keyhive/rules.mk [hunk 8]
```diff
@@ -15,7 +15,7 @@ COMMAND_ENABLE = no         # Commands for debug and configuration
 # Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
 SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
 # if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
```
Comment (fauxpark) [inline keyboards/sofle/keyhive/rules.mk:17]:
```suggestion
```

File: keyboards/sofle/keyhive/readme.md [hunk 9]
```diff
@@ -1,66 +1,53 @@
...
+## Flashing
 
-    make sofle/keyhive:default
-
-Flashing example for this keyboard:
-
-    make sofle/keyhive:default:flash
+Flash using the correct command below (or use QMK Toolbox). These commands can be mixed if, for example, you have an Elite C on the left and a pro micro on the right.
 
 Press reset button on he keyboard when asked.
```
Comment (fauxpark) [inline keyboards/sofle/keyhive/readme.md:39]:
```suggestion
Press reset button on the keyboard when asked.
```

File: keyboards/sofle/keyhive/readme.md [hunk 10]
```diff
@@ -1,66 +1,53 @@
...
 
--   After flashing, in VIA make sure to Import Keymap, which is "sofle VIA keymap.json". This will alow VIA to recognize the updated layout and custom functions. VIA will not auto-recognize the keyboard with this firmware because of the necessary customization.
--   Go to Save+Load to Load Saved Layout. You can import my own layout "sofle VIA layout.json" or just use the Keymap tab to assign your own keys. Having another keyboard connected can be handy for doing this step.
--   It is a good idea to Save Current Layout after you decide on your mapping.
+# Compiling
 
+Make example for this keyboard:
+```sh
+qmk compile -kb sofle/keyhive -km keyhive_via
+```
```
Comment (fauxpark) [inline keyboards/sofle/keyhive/readme.md:33]:
```suggestion

    make sofle/keyhive:default
```

File: keyboards/sofle/keyhive/readme.md [hunk 11]
```diff
@@ -1,66 +1,53 @@
...
 
+```sh
+# for pro micro-based builds
+qmk flash -kb sofle/keyhive -km keyhive_via -bl avrdude-split-left
+qmk flash -kb sofle/keyhive -km keyhive_via -bl avrdude-split-right
+
+# for Elite C or dfu bootloader builds
+qmk flash -kb sofle/keyhive -km keyhive_via -bl dfu-split-left
+qmk flash -kb sofle/keyhive -km keyhive_via -bl dfu-split-right
+```
```
Comment (fauxpark) [inline keyboards/sofle/keyhive/readme.md:51]:
```suggestion
    # for Pro Micro-based builds
    make sofle/keyhive:default:avrdude-split-left
    make sofle/keyhive:default:avrdude-split-right

    # for Elite C or DFU bootloader builds
    make sofle/keyhive:default:dfu-split-left
    make sofle/keyhive:default:dfu-split-right
```

File: keyboards/sofle/keyhive/keyhive.h [hunk 12]
```diff
@@ -21,27 +21,47 @@
...
+        { RE1, RE2, RE3, RE4, RE5, RE6, KC_NO } \
+    }
+
 #define LAYOUT( \
     LA1, LA2, LA3, LA4, LA5, LA6,           RA6, RA5, RA4, RA3, RA2, RA1, \
     LB1, LB2, LB3, LB4, LB5, LB6,           RB6, RB5, RB4, RB3, RB2, RB1, \
     LC1, LC2, LC3, LC4, LC5, LC6,           RC6, RC5, RC4, RC3, RC2, RC1, \
     LD1, LD2, LD3, LD4, LD5, LD6, LE1, RE6, RD6, RD5, RD4, RD3, RD2, RD1, \
-              LE2, LE3, LE4, LE5, LE6, RE5, RE4, RE3, RE2, RE1 			  \
+              LE2, LE3, LE4, LE5, LE6, RE5, RE4, RE3, RE2, RE1            \
```
Comment (fauxpark) [inline keyboards/sofle/keyhive/keyhive.h:51]:
```suggestion
              LE2, LE3, LE4, LE5, LE6, RE5, RE4, RE3, RE2, RE1 \
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/sofle/keyhive/readme.md (hunk 10 above)
Comment (winder) [reply keyboards/sofle/keyhive/readme.md:33]: Are `qmk compile` and `qmk flash` deprecated? It was really convenient that you could run those commands from any directory in the repo.

Comment (fauxpark) [reply keyboards/sofle/keyhive/readme.md:33]: No, but the make commands are what's currently in the template. Eventually that will change.

Comment (winder) [reply keyboards/sofle/keyhive/readme.md:33]: Ok, keeping things consistent until a migration sounds like it would make things easier.

Comment (winder) [reply keyboards/sofle/keyhive/readme.md:33]: I just noticed that you changed the keymap here to default. Almost all of my testing was done with the "keyhive_via" keymap, which was tailor-made for this version of the board. So I wanted to use that as the default.

Comment (fauxpark) [reply keyboards/sofle/keyhive/readme.md]: Again, per the template the "example commands" should use the `default` keymap. They are, after all, examples, not a full guide on compiling a particular board.

Comment (winder) [reply keyboards/sofle/keyhive/readme.md]: Thanks, updated everything to use the default keymap!

**Review by drashna — APPROVED** [review]

File: keyboards/sofle/keyhive/info.json [hunk 13]
```diff
@@ -6,7 +6,16 @@
...
                 {"x":0, "y":2.5}, {"x":1, "y":2.375}, {"x":2, "y":2.125}, {"x":3, "y":2}, {"x":4, "y":2.125}, {"x":5, "y":2.25}, {"x":10.5, "y":2.25}, {"x":11.5, "y":2.125}, {"x":12.5, "y":2}, {"x":13.5, "y":2.125}, {"x":14.5, "y":2.375}, {"x":15.5, "y":2.5},
-                {"x":0, "y":3.5}, {"x":1, "y":3.375}, {"x":2, "y":3.125}, {"x":3, "y":3}, {"x":4, "y":3.125}, {"x":5, "y":3.25}, {"x":6, "y":2.75}, {"x":9.5, "y":2.75}, {"x":10.5, "y":3.25}, {"x":11.5, "y":3.125}, {"x":12.5, "y":3}, {"x":13.5, "y":3.125}, {"x":14.5, "y":3.375}, {"x":15.5, "y":3.5},
+                {"x":0, "y":3.5}, {"x":1, "y":3.375}, {"x":2, "y":3.125}, {"x":3, "y":3}, {"x":4, "y":3.125}, {"x":5, "y":3.25},  {"x":6, "y":2.75}, {"x":9.5, "y":2.75}, {"x":10.5, "y":3.25}, {"x":11.5, "y":3.125}, {"x":12.5, "y":3}, {"x":13.5, "y":3.125}, {"x":14.5, "y":3.375}, {"x":15.5, "y":3.5},
+                {"x":1.5, "y":4.375}, {"x":2.5, "y":4.125}, {"x":3.5, "y":4.15}, {"x":4.5, "y":4.25}, {"x":6, "y":4.25, "h":1.5}, {"x":9.5, "y":4.25, "h":1.5}, {"x":11, "y":4.25}, {"x":12, "y":4.15}, {"x":13, "y":4.125}, {"x":14, "y":4.375}
+            ]
+        }
```
Comment (noroadsleft) [inline keyboards/sofle/keyhive/info.json:12]: Missing comma causes a syntax error.

```suggestion
        },
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

File: keyboards/sofle/keyhive/info.json (hunk 13 above)
Comment (winder) [reply keyboards/sofle/keyhive/info.json]: Thanks, fixed!

File: keyboards/sofle/keyhive/keyhive.h [hunk 14]
```diff
@@ -21,27 +21,47 @@
...
+        { LC1, LC2, LC3, LC4, LC5, LC6, KC_NO }, \
+        { LD1, LD2, LD3, LD4, LD5, LD6, LD7   }, \
+        { LE1, LE2, LE3, LE4, LE5, LE6, KC_NO }, \
+        { RA1, RA2, RA3, RA4, RA5, RA6, KC_NO }, \
+        { RB1, RB2, RB3, RB4, RB5, RB6, RB7   }, \
+        { RC1, RC2, RC3, RC4, RC5, RC6, KC_NO }, \
+        { RD1, RD2, RD3, RD4, RD5, RD6, RD7   }, \
+        { RE1, RE2, RE3, RE4, RE5, RE6, KC_NO } \
+    }
+
```
Comment (noroadsleft) [inline keyboards/sofle/keyhive/keyhive.h:45]: Lines 24-45 should be moved to `keyboards/sofle/keymaps/keyhive_via/keymap.c`, somewhere around Line 37.

File: keyboards/sofle/keyhive/info.json [hunk 15]
```diff
@@ -6,7 +6,16 @@
...
+                {"x":0, "y":1.5}, {"x":1, "y":1.375}, {"x":2, "y":1.125}, {"x":3, "y":1}, {"x":4, "y":1.125}, {"x":5, "y":1.25}, {"x":10.5, "y":1.25}, {"x":11.5, "y":1.125}, {"x":12.5, "y":1}, {"x":13.5, "y":1.125}, {"x":14.5, "y":1.375}, {"x":15.5, "y":1.5},
+                {"x":0, "y":2.5}, {"x":1, "y":2.375}, {"x":2, "y":2.125}, {"x":3, "y":2}, {"x":4, "y":2.125}, {"x":5, "y":2.25}, {"x":10.5, "y":2.25}, {"x":11.5, "y":2.125}, {"x":12.5, "y":2}, {"x":13.5, "y":2.125}, {"x":14.5, "y":2.375}, {"x":15.5, "y":2.5},
+                {"x":0, "y":3.5}, {"x":1, "y":3.375}, {"x":2, "y":3.125}, {"x":3, "y":3}, {"x":4, "y":3.125}, {"x":5, "y":3.25}, {"x":6, "y":0.75}, {"x":6, "y":1.75},  {"x":6, "y":2.75}, {"x":9.5, "y":0.75}, {"x":9.5, "y":1.75}, {"x":9.5, "y":2.75}, {"x":10.5, "y":3.25}, {"x":11.5, "y":3.125}, {"x":12.5, "y":3}, {"x":13.5, "y":3.125}, {"x":14.5, "y":3.375}, {"x":15.5, "y":3.5},
                 {"x":1.5, "y":4.375}, {"x":2.5, "y":4.125}, {"x":3.5, "y":4.15}, {"x":4.5, "y":4.25}, {"x":6, "y":4.25, "h":1.5}, {"x":9.5, "y":4.25, "h":1.5}, {"x":11, "y":4.25}, {"x":12, "y":4.15}, {"x":13, "y":4.125}, {"x":14, "y":4.375}
             ]
```
Comment (noroadsleft) [inline keyboards/sofle/keyhive/info.json:20]:
```suggestion
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

File: keyboards/sofle/keyhive/keyhive.h (hunk 14 above)
Comment (winder) [reply keyboards/sofle/keyhive/keyhive.h]: This makes perfect sense, thanks! It didn't occur to me that this didn't have to be in "<folder-name>.h"

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**